File: day_18/Numbers_With_Same_Consecutive_Differences.py
class Solution:
    def numsSameConsecDiff(self, N: int, K: int) -> List[int]:
        def helper(num, ln):
            last_dig = num % 10
            if ln < N:
                num *= 10
                ln += 1
                if last_dig + K < 10:
                    helper(num + last_dig + K, ln)
                if last_dig - K >= 0:
                    helper(num + last_dig - K, ln)
            else:
                if not self.ans or num != self.ans[-1]:
                    self.ans.append(num)
                return
        self.ans = [] if N > 1 else [0]
        for i in range(1, 10):
            helper(i, 1)
        return self.ans 

File: day_18/Numbers_With_Same_Consecutive_Differences.js
/**
 * Get all non-negative integers of length N such that the absolute
 *  difference between every two consecutive digits is K.
 * @param {number} N
 * @param {number} K
 * @return {number[]}
 */
let numsSameConsecDiff = function (N, K) {
    function dfs(num, ln) {
        // base case
        if (ln === N) {
            // check for not allowed duplicates
            if (!ans || num !== ans[ans.length - 1]) ans.push(num);
            return
        }
        let last_digit = num % 10;
        ln++;
        num *= 10;
        if (last_digit + K < 10) dfs(num + last_digit + K, ln);
        if (last_digit - K >= 0) dfs(num + (last_digit - K), ln);
    }

    let ans = N === 1 ? [0] : [];
    for (let i = 1; i <= 9; i++) {
        dfs(i, 1);
    }
    return ans;
};


console.log(8 * (8 ** 2))
console.log(numsSameConsecDiff(8, 1).length)